cstr/src/lib.rs

```diff
@@ -180,6 +181,77 @@ pub unsafe fn write_to_buffer(message: &str, len: usize, output_ptr: *mut WideCh
     v.len() as u16
 }
 
+///
+/// write_wstring_slice_to_buffer writes the input WideChar slice (wstring) to the output buffer, and returns the number of bytes written
+///
+/// # Safety
+/// This writes to a raw c-pointer, which requires unsafe operations
+pub unsafe fn write_wstring_slice_to_buffer(
+    message: &[WideChar],
+    len: usize,
+    output_ptr: *mut WideChar,
+) -> u16 {
+    let len = std::cmp::min(message.len(), len - 1);
+    let mut v = message[..len].to_vec();
+    v.push(0);
+
+    unsafe {
+        copy_nonoverlapping(v.as_mut_ptr(), output_ptr, len);
+    }
+
+    v.len() as u16
+}
+
+///
+/// write_string_slice_to_buffer writes the input Char slice (string) to the output buffer, and returns the number of bytes written
+///
+/// # Safety
+/// This writes to a raw c-pointer, which requires unsafe operations
+pub unsafe fn write_string_slice_to_buffer(
+    message: &[Char],
+    len: usize,
+    output_ptr: *mut Char,
+) -> u16 {
+    let len = std::cmp::min(message.len(), len - 1);
+    let mut v = message[..len].to_vec();
```

Comment:
We should preallocate this Vec so that the push of 0 doesn't cause a realloc

You can do this using `Vec::with_capacity`:
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity

 and `extend_from_slice`: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extend_from_slice